File: go/test/image_entity_test.go
package sdktest

import (
	"encoding/json"
	"os"
	"path/filepath"
	"runtime"
	"strings"
	"testing"
	"time"

	sdk "github.com/voxgig-sdk/launchpics-sdk/go"
	"github.com/voxgig-sdk/launchpics-sdk/go/core"

	vs "github.com/voxgig-sdk/launchpics-sdk/go/utility/struct"
)

func TestImageEntity(t *testing.T) {
	t.Run("instance", func(t *testing.T) {
		testsdk := sdk.TestSDK(nil, nil)
		ent := testsdk.Image(nil)
		if ent == nil {
			t.Fatal("expected non-nil ImageEntity")
		}
	})

	t.Run("basic", func(t *testing.T) {
		setup := imageBasicSetup(nil)
		// Per-op sdk-test-control.json skip — basic test exercises a flow
		// with multiple ops; skipping any op skips the whole flow.
		_mode := "unit"
		if setup.live {
			_mode = "live"
		}
		for _, _op := range []string{"create", "load", "remove"} {
			if _shouldSkip, _reason := isControlSkipped("entityOp", "image." + _op, _mode); _shouldSkip {
				if _reason == "" {
					_reason = "skipped via sdk-test-control.json"
				}
				t.Skip(_reason)
				return
			}
		}
		// The basic flow consumes synthetic IDs from the fixture. In live mode
		// without an *_ENTID env override, those IDs hit the live API and 4xx.
		if setup.syntheticOnly {
			t.Skip("live entity test uses synthetic IDs from fixture — set LAUNCHPICS_TEST_IMAGE_ENTID JSON to run live")
			return
		}
		client := setup.client

		// CREATE
		imageRef01Ent := client.Image(nil)
		imageRef01Data := core.ToMapAny(vs.GetProp(
			vs.GetPath([]any{"new", "image"}, setup.data), "image_ref01"))

		imageRef01DataResult, err := imageRef01Ent.Create(imageRef01Data, nil)
		if err != nil {
			t.Fatalf("create failed: %v", err)
		}
		imageRef01Data = core.ToMapAny(entityData(imageRef01DataResult))
		if imageRef01Data == nil {
			t.Fatal("expected create result to be a map")
		}
		if imageRef01Data["id"] == nil {
			t.Fatal("expected created entity to have an id")
		}

		// LOAD
		imageRef01MatchDt0 := map[string]any{
			"id": imageRef01Data["id"],
		}
		imageRef01DataDt0Loaded, err := imageRef01Ent.Load(imageRef01MatchDt0, nil)
		if err != nil {
			t.Fatalf("load failed: %v", err)
		}
		imageRef01DataDt0LoadResult := core.ToMapAny(entityData(imageRef01DataDt0Loaded))
		if imageRef01DataDt0LoadResult == nil {
			t.Fatal("expected load result to be a map")
		}
		if imageRef01DataDt0LoadResult["id"] != imageRef01Data["id"] {
			t.Fatal("expected load result id to match")
		}

		// REMOVE
		imageRef01MatchRm0 := map[string]any{
			"id": imageRef01Data["id"],
		}
		_, err = imageRef01Ent.Remove(imageRef01MatchRm0, nil)
		if err != nil {
			t.Fatalf("remove failed: %v", err)
		}

	})
}

func imageBasicSetup(extra map[string]any) *entityTestSetup {
	loadEnvLocal()

	_, filename, _, _ := runtime.Caller(0)
	dir := filepath.Dir(filename)

	entityDataFile := filepath.Join(dir, "..", "..", ".sdk", "test", "entity", "image", "ImageTestData.json")

	entityDataSource, err := os.ReadFile(entityDataFile)
	if err != nil {
		panic("failed to read image test data: " + err.Error())
	}

	var entityData map[string]any
	if err := json.Unmarshal(entityDataSource, &entityData); err != nil {
		panic("failed to parse image test data: " + err.Error())
	}

	options := map[string]any{}
	options["entity"] = entityData["existing"]

	client := sdk.TestSDK(options, extra)

	// Generate idmap via transform, matching TS pattern.
	idmap := vs.Transform(
		[]any{"image01", "image02", "image03"},
		map[string]any{
			"`$PACK`": []any{"", map[string]any{
				"`$KEY`": "`$COPY`",
				"`$VAL`": []any{"`$FORMAT`", "upper", "`$COPY`"},
			}},
		},
	)

	// Detect ENTID env override before envOverride consumes it. When live
	// mode is on without a real override, the basic test runs against synthetic
	// IDs from the fixture and 4xx's. Surface this so the test can skip.
	entidEnvRaw := os.Getenv("LAUNCHPICS_TEST_IMAGE_ENTID")
	idmapOverridden := entidEnvRaw != "" && strings.HasPrefix(strings.TrimSpace(entidEnvRaw), "{")

	env := envOverride(map[string]any{
		"LAUNCHPICS_TEST_IMAGE_ENTID": idmap,
		"LAUNCHPICS_TEST_LIVE":      "FALSE",
		"LAUNCHPICS_TEST_EXPLAIN":   "FALSE",
		"LAUNCHPICS_APIKEY":         "NONE",
	})

	idmapResolved := core.ToMapAny(env["LAUNCHPICS_TEST_IMAGE_ENTID"])
	if idmapResolved == nil {
		idmapResolved = core.ToMapAny(idmap)
	}

	if env["LAUNCHPICS_TEST_LIVE"] == "TRUE" {
		mergedOpts := vs.Merge([]any{
			map[string]any{
				"apikey": env["LAUNCHPICS_APIKEY"],
			},
			extra,
		})
		client = sdk.NewLaunchpicsSDK(core.ToMapAny(mergedOpts))
	}

	live := env["LAUNCHPICS_TEST_LIVE"] == "TRUE"
	return &entityTestSetup{
		client:        client,
		data:          entityData,
		idmap:         idmapResolved,
		env:           env,
		explain:       env["LAUNCHPICS_TEST_EXPLAIN"] == "TRUE",
		live:          live,
		syntheticOnly: live && !idmapOverridden,
		now:           time.Now().UnixMilli(),
	}
}

File: lua/test/image_entity_test.lua
-- Image entity test

local json = require("dkjson")
local vs = require("utility.struct.struct")
local sdk = require("launchpics_sdk")
local helpers = require("core.helpers")
local runner = require("test.runner")

local _test_dir = debug.getinfo(1, "S").source:match("^@(.+/)")  or "./"

describe("ImageEntity", function()
  it("should create instance", function()
    local testsdk = sdk.test(nil, nil)
    local ent = testsdk:Image(nil)
    assert.is_not_nil(ent)
  end)

  it("should run basic flow", function()
    local setup = image_basic_setup(nil)
    -- Per-op sdk-test-control.json skip.
    local _live = setup.live or false
    for _, _op in ipairs({"create", "load", "remove"}) do
      local _should_skip, _reason = runner.is_control_skipped("entityOp", "image." .. _op, _live and "live" or "unit")
      if _should_skip then
        pending(_reason or "skipped via sdk-test-control.json")
        return
      end
    end
    -- The basic flow consumes synthetic IDs from the fixture. In live mode
    -- without an *_ENTID env override, those IDs hit the live API and 4xx.
    if setup.synthetic_only then
      pending("live entity test uses synthetic IDs from fixture — set LAUNCHPICS_TEST_IMAGE_ENTID JSON to run live")
      return
    end
    local client = setup.client

    -- CREATE
    local image_ref01_ent = client:Image(nil)
    local image_ref01_data = helpers.to_map(vs.getprop(
      vs.getpath(setup.data, "new.image"), "image_ref01"))

    local image_ref01_data_result, err = image_ref01_ent:create(image_ref01_data, nil)
    assert.is_nil(err)
    image_ref01_data = helpers.to_map(type(image_ref01_data_result) == 'table' and image_ref01_data_result.data_get and image_ref01_data_result:data_get() or image_ref01_data_result)
    assert.is_not_nil(image_ref01_data)
    assert.is_not_nil(image_ref01_data["id"])

    -- LOAD
    local image_ref01_match_dt0 = {
      id = image_ref01_data["id"],
    }
    local image_ref01_data_dt0_loaded, err = image_ref01_ent:load(image_ref01_match_dt0, nil)
    assert.is_nil(err)
    local image_ref01_data_dt0_load_result = helpers.to_map(type(image_ref01_data_dt0_loaded) == 'table' and image_ref01_data_dt0_loaded.data_get and image_ref01_data_dt0_loaded:data_get() or image_ref01_data_dt0_loaded)
    assert.is_not_nil(image_ref01_data_dt0_load_result)
    assert.are.equal(image_ref01_data_dt0_load_result["id"], image_ref01_data["id"])

    -- REMOVE
    local image_ref01_match_rm0 = {
      id = image_ref01_data["id"],
    }
    local _, err = image_ref01_ent:remove(image_ref01_match_rm0, nil)
    assert.is_nil(err)

  end)
end)

function image_basic_setup(extra)
  runner.load_env_local()

  local entity_data_file = _test_dir .. "../../.sdk/test/entity/image/ImageTestData.json"
  local f = io.open(entity_data_file, "r")
  if f == nil then
    error("failed to read image test data: " .. entity_data_file)
  end
  local entity_data_source = f:read("*a")
  f:close()

  local entity_data = json.decode(entity_data_source)

  local options = {}
  options["entity"] = entity_data["existing"]

  local client = sdk.test(options, extra)

  -- Generate idmap via transform.
  local idmap = vs.transform(
    { "image01", "image02", "image03" },
    {
      ["`$PACK`"] = { "", {
        ["`$KEY`"] = "`$COPY`",
        ["`$VAL`"] = { "`$FORMAT`", "upper", "`$COPY`" },
      }},
    }
  )

  -- Detect ENTID env override before envOverride consumes it. When live
  -- mode is on without a real override, the basic test runs against synthetic
  -- IDs from the fixture and 4xx's. Surface this so the test can skip.
  local entid_env_raw = os.getenv("LAUNCHPICS_TEST_IMAGE_ENTID")
  local idmap_overridden = entid_env_raw ~= nil and entid_env_raw:match("^%s*{") ~= nil

  local env = runner.env_override({
    ["LAUNCHPICS_TEST_IMAGE_ENTID"] = idmap,
    ["LAUNCHPICS_TEST_LIVE"] = "FALSE",
    ["LAUNCHPICS_TEST_EXPLAIN"] = "FALSE",
    ["LAUNCHPICS_APIKEY"] = "NONE",
  })

  local idmap_resolved = helpers.to_map(
    env["LAUNCHPICS_TEST_IMAGE_ENTID"])
  if idmap_resolved == nil then
    idmap_resolved = helpers.to_map(idmap)
  end

  if env["LAUNCHPICS_TEST_LIVE"] == "TRUE" then
    local merged_opts = vs.merge({
      {
        apikey = env["LAUNCHPICS_APIKEY"],
      },
      extra or {},
    })
    client = sdk.new(helpers.to_map(merged_opts))
  end

  local live = env["LAUNCHPICS_TEST_LIVE"] == "TRUE"
  return {
    client = client,
    data = entity_data,
    idmap = idmap_resolved,
    env = env,
    explain = env["LAUNCHPICS_TEST_EXPLAIN"] == "TRUE",
    live = live,
    synthetic_only = live and not idmap_overridden,
    now = os.time() * 1000,
  }
end
